bpf: split off debug options and do not run it in ci

borkmann opened the pull request:
Related: #11175

pchaigno committed:
bpf: Enforce host policies before BPF-based SNAT

This fix is required to get the host policies to work properly with
BPF-backed NodePort services. It is not a sufficient fix, but required
to resolve the 4.19 program size issue in subsequent commits, so sending
now.

Fixes: 88bf291 ("bpf: Enforce host policies for IPv4")
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test: RunsOn419Kernel() helper

Add a small helper that we can use in our CI code.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
cocci: Simplify and extend tail_call.cocci

This commit extends tail_call.cocci to:
- recognize invoke_tailcall_if() calls followed by 'return ret;', as
  used in nodeport.h;
- refactor the code a bit using disjunctions;
- ensure the return statement we are trying to match (e.g., return
  DROP_MISSED_TAIL_CALL or return send_drop_notify()) is not preceeded by
  another return statement.

Fixes: f02bd2a ("cocci: Detect unlogged missed tail calls")
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Remove all of relax_verifier()

relax_verifier() is supposed to help the verifier prune branches and
therefore decrease the reported complexity. On 4.19 however, the call
to relax_verifier in CILIUM_CALL_IPV6_FROM_LXC actually increases the
complexity because it adds instructions in complexity-hot paths.
After a quick check, it doesn't seem to be required for 4.9 either,
therefore remove it entirely.

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: split off debug mode from debugging datapath

In particular running CI with datapath debugging enabled can be harmful
since i) it does not represent what is running in a production environment,
ii) it can hide verifier complexity or instruction size bugs for non-debug
mode which are significantly harder to pin-point and fix 'after the fact'.

Thus, run what would be loaded in a production environment in terms of
BPF code and make that the first-class citizen in CI, not the other way
round. Datapath debugging can still be enabled via CLI through:

  cilium config Debug=Enabled

In the worst case, we can add a new flag to control it via param on agent
start, but given this is mainly just for manual debugging, I left that
out for now.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: remove stale backend affinity deletion from hot-path

It is optional and considered an optimization if you will, however,
i) explicit deletion from there might hit map internal spinlock from
the same hash bucket, and ii) we update the affinity anyway shortly
thereafter or remove it via __lb{4,6}_affinity_backend_id() upon
timeout, so deletion is not necessary. Reduces also instructions.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: constify from_host bool in bpf_host

Given we never change this param, help compiler asserting that its
constant. So it can potentially eliminate unrelated code.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
test: switch up/downgrade tests on 4.19 to disabled debugging

Before this PR, the 4.19 CI pipeline was running probe:

  KubeProxyReplacement: Probe () [HostReachableServices (TCP, UDP), SessionAffinity]

After this PR, it is switched to strict:

  KubeProxyReplacement: Strict (enp0s8) [NodePort (SNAT, 30000-32767, XDP: DISABLED),
                        HostPort, ExternalIPs, HostReachableServices (TCP, UDP),
                        SessionAffinity]

It turns out that this breaks the up/downgrade test on the 1.7.4 image:

  [...]
  2020-06-10T07:22:18.490936835Z level=info msg="Cilium 1.7.4 8ccccd98b 2020-06-03T14:23:45-07:00 go version go1.13.12 linux/amd64" subsys=daemon
  2020-06-10T07:22:18.491423561Z level=debug msg="Load 1-min: 0.16 5-min: 0.23 15min: 0.10" subsys=loadinfo type=background
  2020-06-10T07:22:18.491534313Z level=debug msg="Memory: Total: 7974 Used: 922 (11.56%) Free: 4905 Buffers: 82 Cached: 2064" subsys=loadinfo type=background
  2020-06-10T07:22:18.491684768Z level=debug msg="Swap: Total: 0 Used: 0 (0.00%) Free: 0" subsys=loadinfo type=background
  2020-06-10T07:22:18.511350077Z level=info msg="cilium-envoy  version: 63de0bd958d05d82e2396125dcf6286d92464c56/1.12.3/Modified/RELEASE/BoringSSL" subsys=daemon
  2020-06-10T07:22:18.547219473Z level=info msg="clang (7.0.0) and kernel (4.19.57) versions: OK!" subsys=linux-datapath
  2020-06-10T07:22:18.556721593Z level=info msg="linking environment: OK!" subsys=linux-datapath
  [...]
  2020-06-10T07:22:26.088381054Z level=warning msg="+ tc filter replace dev cilium_vxlan ingress prio 1 handle 1 bpf da obj bpf_overlay.o sec from-overlay" subsys=datapath-loader
  2020-06-10T07:22:26.088412542Z level=warning subsys=datapath-loader
  2020-06-10T07:22:26.08847343Z level=warning msg="Prog section '2/19' rejected: Argument list too long (7)!" subsys=datapath-loader
  2020-06-10T07:22:26.088477964Z level=warning msg=" - Type:         3" subsys=datapath-loader
  2020-06-10T07:22:26.088480166Z level=warning msg=" - Attach Type:  0" subsys=datapath-loader
  2020-06-10T07:22:26.088520672Z level=warning msg=" - Instructions: 4336 (240 over limit)" subsys=datapath-loader
  2020-06-10T07:22:26.088524801Z level=warning msg=" - License:      GPL" subsys=datapath-loader
  [...]

Therefore as a workaround remove datapath BPF debugging output to fix
the complexity issue on 1.7.4. Needs separate addressing on 1.7 via #12018
if we want to bring debugging back.

Related: #12018
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
test: enable/disable debug messages for monitor in runtime

The monitor tests are the only place where we check on DEBUG presence in
the monitor output. Enable BPF datapath debugging there.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
test: fix broken WaitCleanAllTerminatingPodsInNs

Seen this brokeness from a Cilium upgrade test run:

  kubectl delete daemonset --namespace=kube-system cilium
  kubectl delete deployment --namespace=kube-system cilium-operator
  kubectl delete podsecuritypolicy cilium-psp cilium-operator-psp
  kubectl delete daemonset --namespace=kube-system cilium-node-init
  kubectl get pods --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'
  kubectl get pods -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n -n --all-namespaces -o jsonpath='{.items[*].metadata.deletionTimestamp}'

This caused the test suite to wait for an invalid namespace which never
came up. Fix it to move ns into a tmp var.

Fixes: 82df172bfff4 ("test: Add single node conformance test")
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on test/runtime/monitor.go:
```diff
@@ -60,6 +71,17 @@ var _ = Describe("RuntimeMonitorTest", func() {
+		endpoints, err := vm.GetEndpointsIds()
+		Expect(err).Should(BeNil())
+
+		for _, v := range endpoints {
```
Just one issue here which I'll push a patch for: #12037

joestringer committed:
test: Hide host endpoint from runtime tests

The host endpoint introduces some interesting interactions between the
API, the host endpoint datapath configuration management and the
recently proposed autodisabling of debug mode for the host endpoint.
These issues are tracked in more detail here:

https://github.com/cilium/cilium/issues/12037

For now, skip the host endpoint when iterating through endpoints in
runtime tests to simplify working around this issue; coverage is
otherwise the same as for previous releases. We can figure out the right
long-term handling of the host endpoint in RuntimeMonitor tests via the
above issue and unblock PR #11977 with this commit.

Signed-off-by: Joe Stringer <joe@cilium.io>

pchaigno commented on test/helpers/cilium.go:
```diff
@@ -321,7 +321,12 @@ func (s *SSHMeta) GetEndpointsIds() (map[string]string, error) {
+	result := endpoints.KVOutput()
+	delete(result, "")
+
+	return result, nil
```
Why not skip the host endpoint only for `RuntimeMonitorTest` if it's not affecting other tests?

joestringer replied:
Convenience; I started by iterating through several of those conditions and realized I could make a smaller change that would have the intended effect. In the background I was also thinking that there's nothing that's explicitly depending on this state from host endpoint since the host endpoint is brand new.

I think this is worth revisiting when we're not trying to get -rc3 out the door :-) Hence #12037 .

pchaigno commented on test/k8sT/Updates.go:
```diff
@@ -253,9 +253,9 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldHelmChartVers
-			}
+		}
+		if helpers.RunsOnNetNextOr419Kernel() {
+			opts["global.debug.enabled"] = "false"
```
Isn't the same change required for the third `opts["global.debug.enabled"] = "false"` case below (around line 419)?

joestringer replied:
That case around line 419 is just the preflight check, I don't think this flag has any effect there.

pchaigno commented on bpf/lib/common.h:
```diff
@@ -71,12 +71,13 @@
 #define CILIUM_CALL_IPV4_FROM_HOST		22
 #define CILIUM_CALL_IPV6_FROM_HOST		23
-#define CILIUM_CALL_SIZE			24
+#define CILIUM_CALL_IPV6_ENCAP_NODEPORT_NAT	24
```
Nit: I changed that while trying to debug the K8sUpdates test, but would be nice to move `CILIUM_CALL_IPV6_ENCAP_NODEPORT_NAT` back to be right after `CILIUM_CALL_IPV4_ENCAP_NODEPORT_NAT`.
